### PR TITLE
Rework how we cache APKs

### PR DESCRIPTION
**Pull request by jonjohnsonjr**
Fixes https://github.com/chainguard-dev/apko/issues/772
Fixes https://github.com/chainguard-dev/apko/issues/773
Fixes https://github.com/chainguard-dev/apko/issues/780

Rather than caching the entire APK, we will cache the APK sections separately. It is trivial to recombine them using cat to produce the exact original APK, so we don't lose any data.

Doing this gives us two useful things:

1. The process of splitting the APK is relatively expensive, as we have to parse the targz stream to know where to split, exactly. We can amortize this across builds by doing it once per APK.
2. The individual sections have content hashes in the APK data model, so we can use those hashes as keys in the cache (filenames). When we write to the cache, we compute the hashes ourselves, so we get cache invalidation for free.

### CODE REVIEW

**Comment by jonjohnsonjr** [comment]
Cold before:

```
apko publish --keyring-append  --repository-append  --arch amd64    13.19s user 4.11s system 112% cpu 15.437 total
```

image


Cold after:

```
apko publish --keyring-append  --repository-append  --arch amd64    13.59s user 4.08s system 124% cpu 14.236 total
```

image

Notably, we aren't waiting for `fetchPackage` to finish writing the file to disk before we start running ExpandAPK, so the overall execution time drops.

Warm before:

```
apko publish --keyring-append  --repository-append  --arch amd64    13.12s user 4.68s system 223% cpu 7.971 total
```

image


Warm after:

```
apko publish --keyring-append  --repository-append  --arch amd64    10.24s user 2.61s system 187% cpu 6.851 total
```

image


Notably, we no longer pay the price of `ExpandAPK` on cache hits, because we do that once on cache misses and cache the results.

We shave off a second of execution time, and we also do ~20-30% less work gunzipping APKs.

File: pkg/apk/implementation_test.go [hunk 1]
```diff
@@ -48,9 +49,10 @@ QwIDAQAB
 
 var (
 	testPkg = repository.Package{
-		Name:    "alpine-baselayout",
-		Version: "3.2.0-r23",
-		Arch:    testArch,
+		Name:     "alpine-baselayout",
+		Version:  "3.2.0-r23",
+		Arch:     testArch,
+		Checksum: []byte{44, 186, 182, 168, 51, 107, 75, 250, 145, 158, 28, 80, 222, 27, 24, 254, 193, 219, 66, 119},
```
Comment (imjasonh) [inline pkg/apk/implementation_test.go:55]: Can you comment how you generate this? If we change this or add a new test package I don't know how I'd get this value. Just from the APKINDEX?

**Review by imjasonh — APPROVED** [review]

File: pkg/apk/implementation_test.go (hunk 1 above)
Comment (jonjohnsonjr) [reply pkg/apk/implementation_test.go:55]: Great idea.

Comment (jonjohnsonjr) [reply pkg/apk/implementation_test.go:55]: Added some test logging that looks like this:

```
        implementation_test.go:337: did not find cachedPackage("alpine-baselayout") in /var/folders/96/3cdgttts3698hyjg_tydpw5c0000gn/T/TestFetchPackagecache_miss_network_should_fill_cache2563250532/001/https%3A%2F%2Fdl-cdn.alpinelinux.org%2Falpine%2Fv3.16%2Fmain/aarch64/alpine-baselayout-3.2.0-r23: stat /var/folders/96/3cdgttts3698hyjg_tydpw5c0000gn/T/TestFetchPackagecache_miss_network_should_fill_cache2563250532/001/https%3A%2F%2Fdl-cdn.alpinelinux.org%2Falpine%2Fv3.16%2Fmain/aarch64/alpine-baselayout-3.2.0-r23/.ctl.tar.gz: no such file or directory
        implementation_test.go:341:   found "1a3a8e47d2287da6d505d973412cee1ad64bcc17bc5995069e4e932055ecb0c4.dat.tar.gz"
        implementation_test.go:341:   found "2cbab6a8336b4bfa919e1c50de1b18fec1db4277.ctl.tar.gz"
        implementation_test.go:341:   found "2cbab6a8336b4bfa919e1c50de1b18fec1db4277.sig.tar.gz"
```

And a comment that indicates you this is the decoding of `2cbab6a8336b4bfa919e1c50de1b18fec1db4277` from `ctl.tar.gz`.

**Review by imjasonh — APPROVED** [review]

**Pull request merged**